CSS: Add package.json for linting and compiling

m-e-h opened the pull request:
This is obviously a first pass since it hasn't really been discussed or tested beyond my Linux desktop. But I think it could help to guide and keep uniform some of the SCSS/CSS and JS commits.
A `package.json` would give us an agnostic build process. https://github.com/Automattic/_s/issues/1149
It's just `node`, which is what our CSS/JS development tools are built for.
https://github.com/WordPress-Coding-Standards/stylelint-config-wordpress
https://github.com/WordPress-Coding-Standards/eslint-config-wordpress

According to the NPM docs it serves 3 main purposes:
1. It serves as documentation for what packages your project depends on.
2. It allows you to specify the versions of a package that your project can use using semantic versioning rules.
3. Makes your build reproducible which means that its way easier to share with other developers.

This file could also, I think, take the place of and hopefully allow removal of the `.jscsrc` and `.jshintignore` files in the repo. Might be time to switch to ESLint anyway. JSCS was merged into ESLint over a year ago https://core.trac.wordpress.org/ticket/31823

Here's how it works for anybody wanting to submit a CSS pull-request:
From the _s/ root directory in the command line:
first `npm install` _or `yarn`_
then `npm run build:css` _or `yarn build:css`_
And that's it. They've linted the `style.css` for errors, compiled the SCSS, added prefixes according to the specified browser support, and automatically fixed some errors with the remaining errors printed in the terminal for manual fixing.

You can replace `build:css` with any of the other "scripts" commands listed in the file.

grappler commented:
I think we should limit the scope to just CSS and JS for this PR but this will allow us to have a standard way of generating RTL styles (see #1154)  and the POT file (see https://github.com/Automattic/_s/pull/884#issuecomment-312442057).

`stylelint style.css --fix` is still an experimental feature and does not yet automatically fix all of the issues in the `style.css`

If we run CSS Lint on the current `style.css`

```
style.css
   6:277  ⚠  Expected line length to be no more than 80 characters            max-line-length
  14:83   ⚠  Expected line length to be no more than 80 characters            max-line-length
  16:81   ⚠  Expected line length to be no more than 80 characters            max-line-length
  52:23   ⚠  Expected single space after ":" with a single-line declaration   declaration-colon-space-after
 107:15   ⚠  Expected numeric font-weight notation                            font-weight-notation
 247:15   ⚠  Expected numeric font-weight notation                            font-weight-notation
 397:15   ⚠  Expected numeric font-weight notation                            font-weight-notation
 429:23   ⚠  Expected a leading zero                                          number-leading-zero
 433:11   ⚠  Expected a leading zero                                          number-leading-zero
 433:20   ⚠  Expected a leading zero                                          number-leading-zero
 510:9    ⚠  Unexpected named color "royalblue"                               color-named
 514:9    ⚠  Unexpected named color "purple"                                  color-named
 520:9    ⚠  Unexpected named color "midnightblue"                            color-named
 668:15   ⚠  Expected numeric font-weight notation                            font-weight-notation
 800:117  ⚠  Expected line length to be no more than 80 characters            max-line-length
 821:17   ⚠  Expected single space before "{"                                 block-opening-brace-space-before
```

Current result fro eslint
```
/_s/js/customizer.js
   9:1   error  Move the invocation into the parens that contain the function  wrap-iife
  15:5   error  There should be no spaces inside this paren                    space-in-parens
  16:4   error  There should be no spaces inside this paren                    space-in-parens
  20:5   error  There should be no spaces inside this paren                    space-in-parens
  21:4   error  There should be no spaces inside this paren                    space-in-parens
  27:46  error  There should be no spaces inside this paren                    space-in-parens
  30:7   error  There should be no spaces inside this paren                    space-in-parens
  32:46  error  There should be no spaces inside this paren                    space-in-parens
  35:7   error  There should be no spaces inside this paren                    space-in-parens
  36:48  error  There should be no spaces inside this paren                    space-in-parens
  38:7   error  There should be no spaces inside this paren                    space-in-parens
  40:5   error  There should be no spaces inside this paren                    space-in-parens
  41:4   error  There should be no spaces inside this paren                    space-in-parens
  42:3   error  There should be no spaces inside this paren                    space-in-parens

/_s/js/navigation.js
    7:1   error  Move the invocation into the parens that contain the function  wrap-iife
   85:37  error  Expected variable declaration to be on a new line              one-var-declaration-per-line
   90:57  error  There should be no spaces inside this paren                    space-in-parens
  106:3   error  There should be no spaces inside this paren                    space-in-parens

/_s/js/skip-link-focus-fix.js
   8:1  error  Move the invocation into the parens that contain the function  wrap-iife
  31:3  error  There should be no spaces inside this paren                    space-in-parens
```

Before this is merged I would like to see but only if it is decided that we want to do in this direction.
- [ ] CSS Lint and ESLint run in travis
- [ ] Styling issues fixed

m-e-h commented:
A couple of the current lint errors that would need manually fixed are questionable whether or not we can and/or should even "fix" them.
1) `6:277  ⚠  Expected line length to be no more than 80 characters`
This is the theme Description. Can't really be broken into multiple lines that I'm aware of.
2) Several occurrences of `Expected numeric font-weight notation` This is for `font-weight: bold`. Should be fairly safe to change `bold` to `600`, but in rare cases, this may not work correctly with a themes chosen font. Plus, the first occurrence is from Normalize. Probably want to leave that one be.
3) `Unexpected named color "royalblue"` This is for the link styles. Any reason not to use the hex equivalent?

The alternatives to fixing these would be to: 
a) add an exceptions to the Styleint config.
```
  "stylelint": {
    "extends": "stylelint-config-wordpress",
    "font-weight-notation": "numeric"|"named",
  },
```
or 
b) Add an inline comment to the css
```
optgroup {
	font-weight: bold; /* stylelint-disable-line font-weight-notation */
}
```
I'm happy to submit a pull for correctly formatted SCSS files. Just wanted to get thoughts on the above fixes.

grappler commented:
1. True, I justed to wanted to share all of the warning so that we could decide what to do. This may be something we add an exception for so that there is no warning.
2. In the latest update of Normalize they have used `bolder` instead which stops the warning #1155
3. This was added 6 years ago when `_s`  first started and never updated https://github.com/Automattic/_s/commit/a81dc6a810e5ad81d19474c8b7919d5e17cbdf87

If we add an exception it should be added to the config file and not cluter the `style.css`. I don't think we should have an expectation for `font-weight`.

m-e-h commented:
I've updated the `devDependencies` and scripts to use stylefmt for fixing lint errors. It's more mature than Stylelint --fix and supports fixing most rules we're interested in. https://github.com/morishitter/stylefmt#stylelint-rules-that-stylefmt-can-handle
This is the best way I can figure out to keep the formatting automatic.

I also discovered this https://github.com/stylelint/stylelint/tree/master/lib/rules/max-line-length#ignorepattern-regex which I've added to solve the `Description:` and the URL in comments `max-line-length` errors. It isn't working on the description though. Not sure I've got the regex right if anybody has any suggestions.

So after running `npm run css-build` with this updated SCSS branch https://github.com/Automattic/_s/pull/1168
I get this:
```
style.css
   6:277  ⚠  Expected line length to be no more than 80 characters   max-line-length     
 106:15   ⚠  Expected numeric font-weight notation                   font-weight-notation
 248:15   ⚠  Expected numeric font-weight notation                   font-weight-notation
```
Once `normalize.css` gets updated and we figure out the `Description:` regex, `style.css` generation should be pretty flawless.

jrfnl commented on package.json:
```diff
@@ -0,0 +1,46 @@
+    "extends": "stylelint-config-wordpress",
+    "rules": {
+      "max-line-length": [ 80, {
+        "ignorePattern": ["/\b(Description:).*\n/", "/https?:\/\/[0-9,a-z]*.*/"]
```
> It isn't working on the description though. Not sure I've got the regex right if anybody has any suggestions.

Just had a quick look at the regex and it looks fine to me, though the `()` around `Description` should not be needed.
Also, as we know that the `Description` should always be at the start of the line with - if anything - only whitespace before it, you could consider changing the `\b` to the slightly more specific: `^\s*`.
All the same, the regex should work as is as well.

I'm wondering whether `ignorePattern` can actually take an array of patterns or whether the regexes should be merged into one ?

grappler replied:
This `['/(https?://[0-9,a-z]*.*)|(^description\\:.+)|(^tags\\:.+)/i']` should do the trick. Though we may not need it if we you the latest WordPress config.
https://github.com/WordPress-Coding-Standards/stylelint-config-wordpress/pull/151

m-e-h commented:
@jrfnl just tried it without the `https?` pattern and got:
```
   6:277  ⚠  Expected line length to be no more than 80 characters   max-line-length     
  17:81   ⚠  Expected line length to be no more than 80 characters   max-line-length     
 108:15   ⚠  Expected numeric font-weight notation                   font-weight-notation
 250:15   ⚠  Expected numeric font-weight notation                   font-weight-notation
 349:88   ⚠  Expected line length to be no more than 80 characters   max-line-length
```
So it appears that, the `html` pattern was working within the array.

ntwb commented on package.json:
```diff
@@ -0,0 +1,46 @@
+    "css-prefix": "postcss style.css -u autoprefixer -o style.css --no-map",
+    "css-format": "stylefmt style.css",
+    "css-fix": "stylelint style.css --fix",
+    "scss-lint": "stylelint sass/**/*.scss",
```
You need to specify that you are using the SCSS syntax here by adding `--syntax scss` to this command

It should also use the `stylelint-config-wordpress/scss` shared config for SCSS syntax per WPCS for SCSS

ntwb commented on package.json:
```diff
@@ -0,0 +1,46 @@
+      "> 1%",
+      "last 2 versions",
+      "IE 11"
+    ]
```
The browserlist matrix should match that of WordPress' https://make.wordpress.org/design/handbook/design-guide/browser-support/

You can see the above in code form [https://core.trac.wordpress.org/changeset/41062/trunk/Gruntfile.js here]

ntwb commented:
> This is for `font-weight: bold`. Should be fairly safe to change bold to `600`, but in rare cases

`bold` should be replace with `700`, not `600`

ntwb commented:
Lastly, per @davidakennedy comment here https://github.com/Automattic/_s/issues/1149#issuecomment-315438854
> 2. Automate compiling and have it still meet WordPress coding standards.

The two rules you were considering overriding `max-line-length` and `font-weight-notation` should not be as these are part of the [WordPress CSS Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/css)

grappler commented:
@m-e-h I started a PR #1187 to fix the coding standard issues in the JS files reported by PHPCS. I thinking of adding support for ESLint at the same time once `v0.3.0`  for `eslint-config-wordpress` is released.

Would it be OK with you if this PR concentrated on the CSS and mine on the JS?

m-e-h commented:
@grappler No problem! The JS stuff definitely shouldn't be held up by the CSS/Sass can of worms.

@ntwb Thanks for reviewing! I'll clean up this PR once the path for CSS clears up a bit. 
Maybe you could give me your thoughts on using Stylefmt. It seems pretty necessary here for cleaning up the `node-sass` default formatting. Do you see Stylelints `--fix` feature catching up to it's rule coverage anytime soon?

ntwb commented:
> It seems pretty necessary here for cleaning up the node-sass default formatting. 

Hmm,you'll need to look at some of the `node-sass`/`libsass` options I think

I'm guessing this is occurring because the compiled files are then being committed back to the repo, maybe these files should be excluded from the lint checks if they are only ever machine generated.

I'm not that familiar with what file is what in this repo so I'm not sure what the ideal solution here should be, with that said, see the follow up below for a _future_ solution

> Do you see Stylelints --fix feature catching up to it's rule coverage anytime soon?

Eventually, the plan is to basically deprecate stylefmt and have everything moved into stylelint

The v8 release we just released was ~8 months in the making and a few people are taking some time off so things are a little slow at the moment, but now that v8 is out we've a couple of focuses:

One is to add more -fix's to stylelint, the other is to create some _pretty printers_ to support [Prettier](prettier.github.io/prettier/)'s CSS support, so with a combination of both more --fix's and Prettier coming to stylelint I'm sure a solution can be found that will work well for the above scenarios

grappler commented:
@m-e-h As we have a decision in #1149 do you think you could update the PR with the changes @ntwb mentioned and only process the CSS files. Once those are made I can test it more easily.

grappler commented:
@m-e-h Do you think you will have time to update the PR?

m-e-h commented:
@grappler I've updated the PR to get us a little closer.

Do we want a command for linting the SCSS folder? I add the `--syntax scss` flag but I'm not sure how to  extend both the `stylelint-config-wordpress/scss` and `stylelint-config-wordpress` shared configs.

Let me know what else I can change. Don't know that I'm completely clear on what all we want to do here.

grappler commented:
Thanks @m-e-h I was playing around with it yesterday. This is how we could extend both configs.

```json
    "extends": [
        "stylelint-config-wordpress",
        "stylelint-config-wordpress/scss"
    ],
```

I think the first step is to make all of the SCSS pass the linter.
Once it passes the linter we could add as a control point in travis so that pull requests fail if there is a styling issue.
Then we can look at the generating the `style.css` and getting that to pass the linter.
The last step would be to see that travis would automatically generate the style.css

All of this does *not* need to be done in this PR.

ntwb commented on package.json:
```diff
@@ -0,0 +1,38 @@
+    "css-prefix": "postcss style.css -u autoprefixer -o style.css --no-map",
+    "css-format": "stylefmt style.css",
+    "css-fix": "stylelint style.css --fix",
+    "scss-lint": "stylelint sass/**/*.scss --syntax scss"
```
Change this to `stylelint  '**/*.scss' --syntax scss --config node_modules/stylelint-config-wordpress/scss.js`

The glob needs to be quoted and loads the SCSS config from `stylelint-config-wordpress`

m-e-h replied:
Updated. Thanks @ntwb !

grappler commented on package.json:
```diff
@@ -0,0 +1,38 @@
+  "name": "_s",
+  "version": "1.0.0",
+  "scripts": {
+    "css-build": "npm run css-lint && npm run css-compile && npm run css-prefix && npm run css-format && npm run css-fix",
```
I don't think we need to run `css-lint` at the start of the build process. If we were to run a lint it should be `scss-lint`

grappler commented on .gitignore:
```diff
@@ -0,0 +1,3 @@
+# Dependency directories
+node_modules/
+yarn.lock
```
When does `yarn.lock` get created? On my system I am only getting `package-lock.json` generate which we should ignore too.

ntwb replied:
`A `yarn.lock` file is only generated if someone runs `yarn install`

Likewise, a `package-lock.json` is only generated if you run `npm install`

grappler commented on package.json:
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "_s",
+  "version": "1.0.0",
```
We should add a licence and repository field too.

grappler commented on package.json:
```diff
@@ -0,0 +1,38 @@
+    "css-prefix": "postcss style.css -u autoprefixer -o style.css --no-map",
+    "css-format": "stylefmt style.css",
+    "css-fix": "stylelint style.css --fix",
+    "scss-lint": "stylelint 'sass/**/*.scss' --syntax scss --config node_modules/stylelint-config-wordpress/scss.js"
```
Could we reorder the commands so that the linting command and together and the build commands are together?

grappler replied:
I was thinking it may be better to follow the format of `lint:css` and `lint:js` so that we could create the command `lint` to run both.

samikeijonen replied:
That's what many do. +1.

grappler commented:
Thank you @m-e-h for the changes. I have been playing with your changes so that Travis can automatically generate the style.css and commit it. You can see the auto commit here. https://github.com/grappler/_s/commit/175f802aaf84a7f5d9ccb847476c3ad33a3ea758

If we are going to include Sass linting then we should fix the linting issues in the Sass. I am not sure if it would be better to do it in the separate PR where we can discuss the individual issues.

We also need exclude `node_modules`  from PHPCS.

ntwb commented:
> Thank you @m-e-h for the changes. I have been playing with your changes so that Travis can automatically generate the style.css and commit it. You can see the auto commit here. grappler/_s@175f802

I don't think `style.css` should be compiled and committed by Travis CI to be honest, having Travis CI perform 1 part of the workflow that otherwise should be run locally is not ideal IMHO.

If tooling is going to be added, which is the purpose of this pull request then the documentation for underscores in general and the CONTRIBUTING.md should be updated to document how to use these tools for both submitting a pull request that will pass Travis CI's tests and how to use the same tooling for local development for those who fork this repo.

Something like this would be the workflow I'd expect: 

- Lint the SCSS source files using stylelint via `npm run scss-lint`
- Manually fix any SCSS issues
- Compile the CSS via `npm run css-compile`
- Update `style.css` with vendor prefixes using Autoprefixer via `npm run css-prefix `
- Lint the `style.css` file using stylelint via `npm run css-lint`
- Fix any CSS issues using stylelint via `npm run css-fix`
- Fix any CSS issues using stylefmt via `npm run css-format`
- Lint the `style.css` file using stylelint via `npm run css-lint`
- Manually fix any remaining CSS issues **

The npm script `npm run css-build` as it currently stands in this PR omits any stylelint SCSS checks (`npm run scss-lint`), that should be added 

** Neither stylefmt or stylelint contain 100% fixable coverage as such they cannot be relied upon to fix every issue stylelint reports. (I wish stylelint did, but we're not there yet)

grappler commented:
@ntwb Thank you for your feedback. The main discussion of this was in #1149. You will be able to read up why we are moving to travis generating the `style.css` though it is not final as I see some other potential issues with auto commit in workflow.

> Neither stylefmt or stylelint contain 100% fixable coverage as such they cannot be relied upon to fix every issue stylelint reports. (I wish stylelint did, but we're not there yet)

We have some solve this is smaller steps. I am not planning of making a PR for committing the `style.css` travis just yet. There are a few other things that we need to do before. I am happy to help create PR's for stylelint to get those remaining items patched.

ntwb commented:
Thanks, I've read #1149 previously, so yeah, I guess we should walk before we run so to speak 😄 

It would be great to get a hand on working on adding more fixable rules in stylelint @grappler 👍

samikeijonen commented:
What's the status of this PR?

m-e-h commented:
@samikeijonen I'd like to get this thing moving.
Thing is, there are a lot of moving parts, issues, PRs involved.
Can we implement this before we merge the updated normalize.css PR? I think we probably can but for some reason I thought I was sorta waiting on that and a few other SASS/CSS related fixes.

It would be nice to have this in place before any total restructure like what's discussed here https://github.com/Automattic/_s/issues/1215
 
I'm game for whatever. Just let me know what we still need to do to get it merged.

grappler commented:
I would be Ok with this going in before anything else. I think this the most ready PR.

Though I did realize today that the generation of the layouts CSS is not supported but we may change a few things with the new structure.

We can improve the code style for the CSS/Sass and add it to travis later.

samikeijonen commented:
+1 for merging this first.

m-e-h commented:
What other changes do we want to see? Any scripts I should remove?

grappler commented on package.json:
```diff
@@ -0,0 +1,39 @@
+    "compile:css": "node-sass sass/style.scss style.css --output-style expanded --indent-type tab --indent-width 1",
+    "prefix:css": "postcss style.css -u autoprefixer -o style.css --no-map",
+    "lint:css": "stylelint style.css",
+    "lint:css": "stylelint 'sass/**/*.scss' --syntax scss --config node_modules/stylelint-config-wordpress/scss.js",
```
I think this should be `lint:scss`

samikeijonen replied:
Yeah the latter one.

samikeijonen replied:
When I run `npm install` the line `"lint:css": "stylelint style.css",` was removed automatically. Probably because there can't be two same names.

m-e-h replied:
@samikeijonen I've updated this. I had also forgotten to update the names when used within the build command. Should be working correctly now.

m-e-h commented on package.json:
```diff
@@ -0,0 +1,40 @@
+    "stylelint-config-wordpress": "^12.0.0"
+  },
+  "stylelint": {
+    "defaultSeverity": "warning",
```
@grappler Not sure if this is wanted here?
I added it to get rid of errors printed in the terminal when linting didn't return 0 errors.
```
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! _s@1.0.0 fix:css: `stylelint style.css --fix`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the _s@1.0.0 fix:css script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/marty/.npm/_logs/2017-11-15T15_48_26_137Z-debug.log
error Command failed with exit code 2.
```

grappler replied:
I am fine with leaving like this. We can look in the future to change the command to something like `"stylelint '**/*.scss'; exit 0"` if need be.

https://medium.com/@bjankord/how-to-lint-scss-with-stylelint-dc87809a9878

samikeijonen commented:
Thanks for typo fixes!

When I run `npm run build:css` I get weird empty lines before comments in `style.css`. Like this:

````
	z-index: 100000;

	/* Above WP toolbar. */
```

Not a big deal and we should probably use comment before the style in `*.scss` files.

m-e-h commented:
Yep, the way single line comments are currently being used doesn't do well. I don't know how best to handle that. Using `// ` would remove them all together from `style.css`. Not sure if that's OK or not..

samikeijonen commented:
We should leave the comment but I was thinking something like this in the original `scss` file.

```css
/* Above WP toolbar. */
z-index: 100000;

```

m-e-h commented:
There are a couple cases like this:
```
img {
	height: auto; /* Make sure images are scaled correctly. */
	max-width: 100%; /* Adhere to container width. */
}
```
that might not work so well with.

samikeijonen commented:
Is it `fix:css` or `format:css` which are moving around the comments? They should just keep where they are.

m-e-h commented:
This removes the empty line added above the comment:
```
  "stylelint": {
    "defaultSeverity": "warning",
    "extends": "stylelint-config-wordpress",
    "rules": {
        "comment-empty-line-before": "never"
    }
  }
```
but turns the rule off entirely and we probably want it enforced elsewhere.

m-e-h commented:
Just want to clarify what the plan is for moving forward with the `package.json` and the css build process.

Is the plan to merge this first and then use this build process to discuss and clean up the sass folder?

I don't think this file necessarily needs to compile or produce anything in this repo right away but the linting alone will be a good step towards keeping contributors on the same page as the sass gets cleaned up.

grappler commented:
> Is the plan to merge this first and then use this build process to discuss and clean up the sass folder?

That is how I see it. It is better to have smaller PR's then one big one.

samikeijonen commented:
We can safely merge this because at first state we're not building `style.css` with it. 

I think there is roadmap somewhere in which order we should continue. If not, let's create one.

samikeijonen commented:
Any roadmap for merging this? I'll ping @davidakennedy just in case :)

timurc commented:
Hey! I was researching bare themes to start from, and this seems really one of the best to me. I would love to have a build process build in, this looks like a good start.

Ismail-elkorchi commented:
Thank you  @m-e-h for the PR, I was a great inspiration for me.

I'm going to close this PR as linting and compiling tools were added in https://github.com/Automattic/_s/pull/1386, and https://github.com/Automattic/_s/pull/1391.
